### ISSUE TITLE
Skinning use cases and examples

### ISSUE THREAD

**Issue by donmccurdy** [KhronosGroup/glTF#1403]
We currently do not have example assets for many of the less-common situations involving skinning, which has prompted questions about whether they are valid and how to implement them.

- [ ] Joint hierarchy does has no common root. 
  - As discussed, the `skeleton` property should probably be removed, and can already be ignored by implementations. In addition to that, we should have an example asset for which no common root is possible (e.g. bones form two subtrees starting from the scene root).
- [ ] Joint hierarchy is intermixed with non-joint nodes. 
  - Such as having a mesh attached to a joint, or having empty non-joint nodes between joints of the same skeleton. (https://github.com/donmccurdy/three-gltf-viewer/issues/101)
- [ ] Animation transforming parents of joints (should transform skinned mesh) or parents of skinned mesh (should have no effect). (#1195)
- [ ] Multiple meshes sharing a single skin.
- [ ] Many (>4) joint vertex influences.

In each case, if we find that an example cannot be consistently implemented we may tighten the spec accordingly.

**Comment by donmccurdy** [comment]
@bghgary would some of these be reasonably within scope for glTF-Asset-Generator to output? I can start making some one-off models in Blender if not.

**Comment by donmccurdy** [comment]
Added another edge case above, "Node references a `skin` but not a `mesh`.". See https://github.com/KhronosGroup/glTF/issues/1448.

**Comment by lexaknyazev** [comment]
@donmccurdy This isn't allowed by schema:
https://github.com/KhronosGroup/glTF/blob/6612de29c8d182ba8d374491ce43d38a6be0e07f/specification/2.0/schema/node.schema.json#L85-L88

**Comment by donmccurdy** [comment]
Ah good to know. Removed the addition above.

**Comment by donmccurdy** [comment]
Added `A subset of joints in one skin are reused in another skin.` above, per https://github.com/KhronosGroup/glTF/issues/1454.

**Comment by donmccurdy** [comment]
Thanks to @kcoley, @stevk, and @bghgary, for adding [new test cases](https://github.com/KhronosGroup/glTF-Asset-Generator/blob/master/Output/Animation_Skin/README.md) in glTF-Asset-Generator, with [more on the way](https://github.com/KhronosGroup/glTF-Asset-Generator/pull/465). 🎉

From those that are available now, I've filled in three of the "example" links above:

> Joint hierarchy is intermixed with non-joint nodes. (donmccurdy/three-gltf-viewer#101) 

This works fine for all of the three implementations tested (BabylonJS, ThreeJS, PlayCanvas). We may want to test a native engine like Unreal or Unity as well, but it seems like this case can probably be allowed, and should possibly be mentioned as an implementation note. It is a known use case that engines may attach objects to bones after loading, and I see no reason to disallow it in the model itself.

> Animation transforming parents of joints (works) or parents of skinned mesh (no effect). (#1195) 

Works in all three, and the spec is already clear on this case. 

> Some or all joints are detached from the scene hierarchy (#1383) 

Works only in BabylonJS. I'm not sure whether this should be allowed.

**Comment by zellski** [comment]
Pragmatic question -- for an exporter like FBX2glTF, do you think it's reasonable to cater to the loaders that have issues with your example #1 (joints form a strict subset of the relevant sub-scene-graph) by simply adding the inactive but tree-connecting nodes as joints in the skin? It definitely solves the issue I'm working on, but I'm not clever enough to figure out if it's likely to cause other issues on some esoteric models.

**Comment by bghgary** [comment]
I've update the first comment with the latest examples from asset generator v0.6. The only one that is missing is the >4 joints one because no engines support it right now.

**Comment by bghgary** [comment]
> adding the inactive but tree-connecting nodes as joints in the skin?

Would you also add weights and inverse bind matrices for them? I think the down side is that it costs more space?

**Comment by zellski** [comment]
> > adding the inactive but tree-connecting nodes as joints in the skin?
> 
> Would you also add weights and inverse bind matrices for them? I think the down side is that it costs more space?

I would, and it does, which is irritating. Still, it's per-node waste rather than per-vertex waste, and on the models I just tested on, the size increase is a small fraction of a percent. I can live with that (and if others hate it it's easy to add command line flags). I'm more worried about unintended side effects I hadn't though of –– to say I'm weak on skinning is an understatement.

**Comment by bghgary** [comment]
I don't call myself an expert on skinning either, but I don't think there will be any side effects. BTW, which viewers don't work correctly without this?

**Comment by zellski** [comment]
> BTW, which viewers don't work correctly without this?

Windows 3D Viewer stumbles like this: 
![image](https://user-images.githubusercontent.com/714167/48520279-db450380-e824-11e8-9c45-bcb2e63ffafb.png)

If we peek at the bones we see unnaturally many of them are rooted in the same spot on the lumbar spine. The misplaced limbs (head, left arm, right arm) all sit on nodes that don't don't have FbxSkin deformers in the original FBX, so they don't become joints. Maybe the loader assumes a tree, interprets this skeleton as four disjoint trees, and roots them all in the common shared ancestor node?

![image](https://user-images.githubusercontent.com/714167/48520448-7807a100-e825-11e8-89bf-4d672c6aee61.png)

After the "fix" all is well (and animations run correctly too, so WEIGHTS_0 and JOINTS_0 appear to be undisturbed).
![image](https://user-images.githubusercontent.com/714167/48520739-91f5b380-e826-11e8-9023-3a8bd0e74e0e.png)

Yikes, that was a lot of large images. Sorry to spam.

**Comment by bghgary** [comment]
Yes, confirmed with 3D viewer. This will get fixed as we require that the app pass all asset generator tests.

**Comment by zen85** [comment]
I also have the problem that constraints on a mesh to another mesh or parenting does not really work. i get it to the point where the mesh gets positioned correctly once in the beginning but does not move along when iE the hand of the model moves.

i am relativly new to this: do i understand you correctly that this is a known issue of the exporter (in my case Blender) but you need testdata to solve that issue? because i would happily provide you with that if you tell me what you need and i can safe some of your precious time since it seems you are busy on so many forums answering our questions. can i contribute/support you at this somehow? 

i just want to attach supersimple eyebrows on my robotface with morph-targets to get some kind of expression in that face. or is there another route to get that working? i want to achive something similar to that: https://threejs.org/examples/?q=morph#webgl_animation_skinning_morph

**Comment by donmccurdy** [comment]
@zen85 it sounds like you're describing a bug in a Blender exporter? If you haven't already, could you file a bug on the exporter you're using? If it's the one that ships with Blender, glTF-Blender-IO, we're tracking a related issue here: https://github.com/KhronosGroup/glTF-Blender-IO/issues/75.

If you're able to share an example model and attach it to the issue above, that would be helpful. Thanks!
